app/src/androidTest/java/com/google/jetpackcamera/utils/AppTestUtil.kt

```diff
@@ -15,12 +15,124 @@
+private val APP_REQUIRED_PERMISSIONS: List<String> = buildList {
     add(android.Manifest.permission.CAMERA)
     add(android.Manifest.permission.RECORD_AUDIO)
     if (Build.VERSION.SDK_INT <= 28) {
         add(android.Manifest.permission.WRITE_EXTERNAL_STORAGE)
     }
 }
+
+val TEST_REQUIRED_PERMISSIONS: List<String> = buildList {
+    addAll(APP_REQUIRED_PERMISSIONS)
+    if (Build.VERSION.SDK_INT >= 33) {
+        add(android.Manifest.permission.READ_MEDIA_IMAGES)
+        add(android.Manifest.permission.READ_MEDIA_VIDEO)
+    }
+}
+
+fun mediaStoreInsertedFlow(
+    mediaUri: Uri,
+    instrumentation: Instrumentation,
+    filePrefix: String = ""
+): Flow<Pair<String, Uri>> = with(instrumentation.targetContext.contentResolver) {
+    // Creates a map of the display names and corresponding URIs for all files contained within
+    // the URI argument. If the URI is a single file, the map will contain a single file.
+    // On API 29+, this will also only return files that are not "pending". Pending files
+    // have not yet been fully written.
+    fun queryWrittenFiles(uri: Uri): Map<String, Uri> {
+        return buildMap {
+            query(
+                uri,
+                buildList {
+                    add(BaseColumns._ID)
+                    add(MediaStore.MediaColumns.DISPLAY_NAME)
+                    if (Build.VERSION.SDK_INT >= 29) {
+                        add(MediaStore.MediaColumns.IS_PENDING)
+                    }
+                }.toTypedArray(),
+                null,
+                null,
+                null
+            )?.use { cursor: Cursor ->
+                cursor.moveToFirst()
+                val idCol = cursor.getColumnIndex(BaseColumns._ID)
+                val displayNameCol = cursor.getColumnIndex(MediaStore.MediaColumns.DISPLAY_NAME)
+
+                while (!cursor.isAfterLast) {
+                    val id = cursor.getLong(idCol)
+                    val displayName = cursor.getString(displayNameCol)
+                    val isPending = if (Build.VERSION.SDK_INT >= 29) {
+                        cursor.getInt(cursor.getColumnIndex(MediaStore.MediaColumns.IS_PENDING))
+                    } else {
+                        // On devices pre-API 29, we don't have an is_pending column, so never
+                        // say that the file is pending
+                        0
+                    }
+                    if (isPending == 0 &&
+                        (filePrefix.isEmpty() || displayName.startsWith(filePrefix))
+                    ) {
+                        // Construct URI for a single file
+                        val outputUri = if (uri.lastPathSegment?.equals("$id") == false) {
+                            uri.buildUpon().appendPath("$id").build()
+                        } else {
+                            uri
+                        }
+                        put(displayName, outputUri)
+                    }
+                    cursor.moveToNext()
```

Comment:
just making sure, this cursor doesn't need to be closed right?

Reply:
The cursor is automatically closed by virtue of accessing it via a `use {}` block (on line 69).